src/rgw/cls_fifo_legacy.cc

```diff
@@ -1358,7 +1355,9 @@ int FIFO::push(const DoutPrefixProvider *dpp, const std::vector<cb::list>& data_
   if (need_new_head) {
     ldpp_dout(dpp, 20) << __PRETTY_FUNCTION__ << ":" << __LINE__
 		   << " need new head tid=" << tid << dendl;
+    std::unique_lock l2(m2);
     r = _prepare_new_head(dpp, head_part_num + 1, tid, y);
+    l2.unlock();
```

Comment:
i'm concerned about holding locks over i/o here, especially given the interaction with `optional_yield`. when we do have a yield context, this suspended coroutine could resume on any other thread. calling `unlock()` from a different thread is undefined behavior - see https://en.cppreference.com/w/cpp/thread/mutex/unlock:
> The mutex must be locked by the current thread of execution, otherwise, the behavior is undefined.

Reply:
True, unlock from a different thread will be problematic.
However in the codebase, the yield passed is a null_yield , so does coroutine even do a context switch ?
Second, currently there is lock used  in cls_fifo_legacy.cc for other use cases, so then lock-unlock would have been issue there too ?

Reply:
> However in the codebase, the yield passed is a null_yield , so does coroutine even do a context switch ?

thanks, that would explain why it doesn't crash. but we will eventually want to cover all of these cases with a real yield context

> Second, currently there is lock used in cls_fifo_legacy.cc for other use cases, so then lock-unlock would have been issue there too ?

the mutex is fine so long as it's not held over a suspension point

Reply:
Thanks for the update, regarding the mutex hold over a suspension point..
in current code base the mutex is held intermittently over following functions (_prepare_new_head --> _prepare_new_part --> update_meta --> apply_update --> process_journal --> create_part -> update_meta --> apply_update)  and i was proposing to hold 1 single lock across all of these functions.
so in both scenarios, if the suspension point is within the window of execution of any of these functions, we will face the issue in future when the null_yield is replaced with real_yield ?
Also with coroutine we can control the suspension point ? so if we control the suspension point to not be within the push window we should be good ?

Reply:
the suspend happens inside of `rgw_rados_operate()`. given a null_yield, that makes a synchronous call to librados, blocking until we get the osd's response. given a yield context, we can suspend instead of blocking, so the current thread of execution can run other coroutines in the meantime

Reply:
Thanks for the explanation, so that make sense now, having single lock across all function will make encapsulate rgw_rados_operate() as well, and we do not need that coz in future when we pass the yield the unlock could happen on some other thread. So only option we have is to keep intermittent and small locks across all functions !!

what about another minor code change as part of this PR, to move the code to increment the version.ver inside the cls_fifo_type.h instead of incrementing it separately for osd and rgw ?
Will remove the lock changes from here.